packages/fxa-shared/metrics/validate.ts

```diff
@@ -0,0 +1,379 @@
+            limitOrDefault: error.limitOrDefault,
+          });
+          scope.setTag('type', 'metrics');
+          this.sentry.captureException(error);
+        });
+      }
+    }
+  }
+}
+
+/** Optional configuration for Metric Validator */
+export type MetricValidatorOpts = {
+  /** Maximum allowed time for an event offset in milliseconds. Generally defaults to 2 days. */
+  maxEventOffset: number;
+  /** Validates a utm_* query param */
+  isUtmValid?: (val: string) => boolean;
+  /** Validates a device id. */
+  isDeviceIdValid?: (val: string) => boolean;
+};
+
+/** Checks and cleans metrics data. */
+export class MetricValidator {
+  protected readonly maxEventOffset = Number.MAX_SAFE_INTEGER;
+  protected readonly isUtmValid = (x: string) => UTM_REGEX.test(x);
+  protected readonly isDeviceIdValid = (x: string) => DEVICE_ID_REGEX.test(x);
+
+  /**
+   * Creates a new instance
+   * @param errorReporter reports issues to sentry and tracks error counts
+   * @param opts optional configuration
+   */
+  constructor(
+    public readonly errorReporter: MetricErrorReporter,
+    opts: MetricValidatorOpts
+  ) {
+    if (opts.maxEventOffset != null) {
+      this.maxEventOffset = opts.maxEventOffset;
+    }
+    if (opts.isDeviceIdValid != null) {
+      this.isUtmValid = opts.isDeviceIdValid;
+    }
+    if (opts.isUtmValid != null) {
+      this.isUtmValid = opts.isUtmValid;
+    }
+  }
+
+  /**
+   * Ensures the device adheres to a hex string of length 32 or 'none'.
+   *
+   * Coercion of data will be reported but NOT counted as a critical error.
+   */
+  public sanitizeDeviceId(data: Pick<CheckedMetrics, 'deviceId'>) {
+    if (this.isDeviceIdValid(data.deviceId || '')) {
+      return data;
+    }
+
+    data.deviceId = NOT_REPORTED_VALUE;
+
+    // The device is not valid, let's try to figure out where this is coming from, by
+    // capturing a sentry error which could uncover the call stack that resulted in
+    // this state. We won't count this as critical, however, because the NOT_REPORTED_VALUE
+    // won't be rejected by the API.
+    this.errorReporter.report(
+      new InvalidMetricError(
+        'deviceId',
+        'invalid value',
+        false,
+        data.deviceId,
+        NOT_REPORTED_VALUE
+      )
+    );
+
+    return data;
+  }
+
+  /**
+   * This will make sure duration is a sane value.
+   *
+   * Coercion of data will be reported and counted as a critical error.
+   */
+  public sanitizeDuration(data: Pick<CheckedMetrics, 'duration'>) {
+    data.duration = this.clamp(
+      'duration',
+      data.duration,
+      0,
+      MAX_SAFE_INTEGER,
+      0
+    );
+
+    return data;
+  }
+
+  /**
+   * Make sure navigation timings are sane values.
+   *
+   * Coercion of data will be reported and counted as a critical error.
+   */
+  public sanitizeNavigationTiming(
+    data: Pick<CheckedMetrics, 'navigationTiming'>
+  ) {
+    // Navigation timing
+    if (data.navigationTiming == null) {
+      return;
```

Comment:
Do we want to `return data` here too for a consistent return type?

Reply:
We will no longer return 'data' from these sanitize methods.